src/backend/tls.rs

```diff
@@ -0,0 +1,247 @@
+                x509.set_pubkey(&ca_key)?;
+                let mut serial = BigNum::new()?;
+                serial.rand(64, MsbOption::MAYBE_ZERO, false)?;
+                x509.set_serial_number(Asn1Integer::from_bn(&serial)?.as_ref())?;
+                x509.set_version(
+                    3 - 1, // zero-indexed
+                )?;
+                let ctx = x509.x509v3_context(None, Some(&conf));
+                let exts = [
+                    BasicConstraints::new().critical().ca().build()?,
+                    SubjectKeyIdentifier::new().build(&ctx)?,
+                    AuthorityKeyIdentifier::new()
+                        .issuer(false)
+                        .keyid(false)
+                        .build(&ctx)?,
+                    KeyUsage::new()
+                        .critical()
+                        .digital_signature()
+                        .key_cert_sign()
+                        .crl_sign()
+                        .build()?,
+                ];
+                for ext in exts {
+                    x509.append_extension(ext)?;
+                }
+                x509.sign(&ca_key, MessageDigest::sha256())?;
+                Ok(x509)
+            })
+            .unwrap()
+            .build();
+        Self { ca_key, ca_cert }
+    }
+
+    pub async fn get_or_create_k8s_certificate(
+        client: &stackable_operator::client::Client,
+    ) -> Self {
+        let k8s_secret_name = "secret-provisioner-ca";
+        let k8s_ns = "default";
+        let existing_secret = client.get::<Secret>(k8s_secret_name, Some(k8s_ns)).await;
+        match existing_secret {
+            Ok(ca_secret) => {
+                let ca_data = ca_secret.data.unwrap_or_default();
+                Self {
+                    ca_key: PKey::private_key_from_pem(&ca_data.get("ca.key").unwrap().0).unwrap(),
+                    ca_cert: X509::from_pem(&ca_data.get("ca.crt").unwrap().0).unwrap(),
+                }
+            }
+            Err(_) => {
+                // Failed to get existing cert, try to create a new self-signed one
+                let ca = Self::new_self_signed();
+                // Use create rather than apply so that we crash and retry on conflicts (to avoid creating spurious certs that we throw away immediately)
+                client
+                    .create(&Secret {
+                        metadata: ObjectMetaBuilder::new()
+                            .namespace(k8s_ns)
+                            .name(k8s_secret_name)
+                            .build(),
+                        data: Some(
+                            [
+                                (
+                                    "ca.key".to_string(),
+                                    ByteString(ca.ca_key.private_key_to_pem_pkcs8().unwrap()),
+                                ),
+                                (
+                                    "ca.crt".to_string(),
```

Comment:
Are these unwraps safe? If so, can we replace them with .expect() ?

Reply:
They are safe in the sense that they don't depend on any user-provided data.

I agree that `expect` might be clearer, but #21 turns them into regular `Error`s (since we start pulling in the data from the `SecretClass` rather than hard-coding it).

Reply:
Just noticed that some of my comments were addressed in #21 yes, came back here to mention that, but you were too quick for me!